### PR TITLE
[euvr] Subscription/Billing updates

### PR DESCRIPTION
**Pull request by vincentsalucci**
## Type of change

- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Code changes
- **organization-billing.component.html**: Copied over existing `user-billing.component.html` to maintain backwards compatibility for organizational settings (to be updated in the near future)
- **organization-billing.component.ts**: Removed base class and gave this component all the capabilities necessary to maintain backwards compatibility for organizational settings (to be updated in the near future)
- **oss-routing.module.ts**: Updated routing to match new design pattern
- **oss.module.ts**: Added newly created components
- **payment-method.component.html**: Created. Took useful pieces from the `user-billing.component`
- **payment-method.component.ts**: Created. Took useful pieces from the `user-billing.component`
- **premium.component.html**: Updated to match new design. Updated with new `CL` components where applicable
- **premium.component.ts**: Updated route to redirect when a user already has premium
- **settings.component.html**: Removed old navigation items and replaced with single `Subscription` option
- **settings.component.ts**: Added additional logic that will handle hiding the `Subscription` option under certain conditions.
- **subscription.component.html**: Created container for routing to selected tabs within the `Subscription` navigation option
- **subscription.component.ts**: Created. Handles supply of any logic needed for adjusting routes
- **subscription.module.ts**: Created. Handles lazy loading of child routes for the `Subscription` navigation option
- **user-billing.component.html**: Removed all `payment-method` related UI and updated with new `CL` components where applicable
- **user-billing..component.ts**: Removed all `payment-method` variables/logic/methods
- **user-subscription.component.html**: Adjusted for updated designs.  uUpdated with new `CL` components where applicable
- **user-subscription.component.ts**: Added helper method for title change. Updated navigation scheme on redirect.
- **messages.json**: Added `billingHistory`

## Screenshots
**Hidden Subscription due to proper conditions**
4-security-keys
5-sub-sub
6-sub-payment
7-sub-billing
8-sub-self-host

## Testing requirements
- Subscription flows for both cloud and self-hosted
- Organizational billing (should work as it currently does - need regression)

## Dependency
- [jslib - Separate Billing API](https://github.com/bitwarden/jslib/pull/750)

## Before you submit
- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/app/settings/user-subscription.component.ts
Comment (eliykat) [inline src/app/settings/user-subscription.component.ts:215]: Is this in the wireframes? I don't understand why we'd change the title this way.

Comment (eliykat) [inline src/app/settings/user-subscription.component.ts:215]: Is this in the wireframes? I don't understand why we'd change the title this way.

**Review by eliykat — CHANGES_REQUESTED** [review]
Nice work! Just some smaller comments.

**Comment by addisonbeck** [comment]
It looks like this PR is targeting `Feature/EndUserVaultRefresh`. There were some concerns raised from others about pascal case breaking some git clients. Can we change this to target `feature/endUserVaultRefresh`?

**Comment by vincentsalucci** [comment]
> It looks like this PR is targeting `Feature/EndUserVaultRefresh`. There were some concerns raised from others about pascal case breaking some git clients. Can we change this to target `feature/endUserVaultRefresh`?

@addisonbeck - I saw that was mentioned in chat but had no idea we officially switched to `feature/endUserVaultRefresh`. I've also merged code into the now obsolete branch. I will cherry-pick to the new branch and try to re-point this one. Maybe that's why my build is broken /shrug

**Comment by eliykat** [comment]
@vincentsalucci I'm making changes to these files in `master` for EC-74 - here's the diff so far: https://github.com/bitwarden/web/compare/feature/msp-hide-billing

This is going to cause a merge conflict as soon as we merge `feature/endUserVaultRefresh` into `master`, which will not have an obvious resolution.

What do you think about creating the standalone `organization-billing.component` in `master` first, as a separate PR? Then you can merge that change through into these feature branches, and I can safely make EC-74 changes to the org billing component without affecting EUVR work.

File: src/app/organizations/settings/organization-billing.component.ts
Comment (eliykat) [inline src/app/organizations/settings/organization-billing.component.ts:18]: Can you add a `// TODO` somewhere in this file, noting that it should be combined with `payment-method.component` when we do the org admin refresh? I think it's reasonable to just copy + paste the logic for now so that we don't go out of scope, but I want to make sure it's reconsidered in the next cycle.

**Comment by vincentsalucci** [comment]
> @vincentsalucci I'm making changes to these files in `master` for EC-74 - here's the diff so far: https://github.com/bitwarden/web/compare/feature/msp-hide-billing
> 
> This is going to cause a merge conflict as soon as we merge `feature/endUserVaultRefresh` into `master`, which will not have an obvious resolution.
> 
> What do you think about creating the standalone `organization-billing.component` in `master` first, as a separate PR? Then you can merge that change through into these feature branches, and I can safely make EC-74 changes to the org billing component without affecting EUVR work.

@eliykat I will work on this asap to try and get it to you soon ™️

File: src/app/settings/settings.component.ts
Comment (vincentsalucci) [reply src/app/settings/settings.component.ts:56]: Ah, I see - I did the stretch goal task code after I took the original screenshots. I apologize and see how that's confusing :D  Will add a screenshot of the `Subscription` vertical navigation tab being hidden as expected 👍

File: src/app/settings/settings.component.ts
Comment (vincentsalucci) [reply src/app/settings/settings.component.ts:9]: It is! Just a bit confusing because this is actually a web specific `StateService` that allows for use of additional abstracted methods. It extends our `jslib` `StateService` abstraction within.

File: src/app/organizations/settings/organization-billing.component.ts
Comment (vincentsalucci) [reply src/app/organizations/settings/organization-billing.component.ts]: I'm not sure I follow - the org admin refresh will follow the same pattern as the current refresh and separate billing/invoice and payment method/balance.

File: src/app/settings/premium.component.html
Comment (vincentsalucci) [reply src/app/settings/premium.component.html]: Updated!

File: src/app/settings/user-billing.component.html
Comment (vincentsalucci) [reply src/app/settings/user-billing.component.html:1]: Updated to `user-billing-history`!

**Comment by eliykat** [comment]
> > @vincentsalucci I'm making changes to these files in `master` for EC-74 - here's the diff so far: https://github.com/bitwarden/web/compare/feature/msp-hide-billing
> > This is going to cause a merge conflict as soon as we merge `feature/endUserVaultRefresh` into `master`, which will not have an obvious resolution.
> > What do you think about creating the standalone `organization-billing.component` in `master` first, as a separate PR? Then you can merge that change through into these feature branches, and I can safely make EC-74 changes to the org billing component without affecting EUVR work.
> 
> @eliykat I will work on this asap to try and get it to you soon ™️

Sorry, turns out I won't actually be touching this component, so you don't need to worry about this extra work. Please proceed with the original PR as you were :)

File: src/app/organizations/settings/organization-billing.component.ts
Comment (eliykat) [reply src/app/organizations/settings/organization-billing.component.ts]: Basically, there's a lot of copy + pasted logic duplicated between `organization-billing.component.ts` and `user-billing.component.ts` as a result of these changes. My understanding is that this is temporary to preserve the existing behaviour of the org pages. My thinking is, we want some flag here to make sure this duplicated code doesn't hang around. Otherwise it should be moved to a base class so it can be shared between components.

Comment (vincentsalucci) [reply src/app/organizations/settings/organization-billing.component.ts:18]: It is only temporary and there will be no duplicated code once the org admin refresh takes place as the `organization-billing.component.ts` will be refactored to extend the `user-billing-history.component.ts` as it exists in production today. Will add a `TODO` as suggested as another reminder!

File: src/app/settings/user-subscription.component.ts
Comment (vincentsalucci) [reply src/app/settings/user-subscription.component.ts:215]: Good catch: It wasn't originally but the wireframes do have a header now. For the `Self Hosted` flow it shows `Subscription` instead of the current `Premium Membership` - I will double check with design today, since I technically did the headers before they were added into the wireframes. 👍

File: src/app/settings/user-billing.component.html
Comment (vincentsalucci) [reply src/app/settings/user-billing.component.html]: Updated!

File: src/app/settings/user-subscription.component.ts
Comment (vincentsalucci) [reply src/app/settings/user-subscription.component.ts:215]: I talked with design and the wireframes have been updated to reflect what is currently coded. This should be good to review 👍

File: src/app/organizations/settings/organization-billing.component.ts
Comment (eliykat) [reply src/app/organizations/settings/organization-billing.component.ts:18]: Yes I agree - we're on the same page, just wanted a TODO in there as a reminder, once we've moved on and dumped all of this out of our heads :D

**Review by eliykat — COMMENTED** [review]
2 differences I noticed from the wireframes - not sure if it's intentional:

1. The wireframe has the Premium Membership tab laid out more vertically:

Screen Shot 2022-05-05 at 3 08 43 pm

Compare to the current layout:

Screen Shot 2022-05-05 at 3 08 36 pm

2. You have a divider line under your headings, whereas the Figma doesn't - also visible in the screenshots above. The reason I noticed is that it starts to carve up the page a little especially in Billing History:

Screen Shot 2022-05-05 at 3 12 32 pm

If design says it's fine then it's fine, just checking.

File: src/app/settings/user-subscription.component.html
Comment (eliykat) [inline src/app/settings/user-subscription.component.html:38]: Any reason for removing this? We should always specify `type` for buttons, and `type="button"` is a good default. Otherwise it can default to a submit button and cause unexpected behaviour.

Comment (eliykat) [reply src/app/settings/user-subscription.component.html:38]: Same comment for all other buttons on this page where you've removed the `type` attribute.

**Review by eliykat — CHANGES_REQUESTED** [review]

File: src/app/settings/user-subscription.component.html
Comment (Hinton) [reply src/app/settings/user-subscription.component.html:38]: We should omit `type="button"` on `button` elements unless it's in a form, in which case we need type to not make it a submit button.

Comment (eliykat) [reply src/app/settings/user-subscription.component.html:38]: That seems reasonable, it is form behaviour that I was concerned about. I couldn't find the PR you refer to, what's the accessibility reason? Is it read out by screenreaders or something?

Comment (Hinton) [reply src/app/settings/user-subscription.component.html:38]: Seems to be my bad. I had confused `role` and `type`. Let's keep `type="button"` to be safe.

Comment (vincentsalucci) [reply src/app/settings/user-subscription.component.html:38]: Added `type="button"`back to the button declarations

**Comment by vincentsalucci** [comment]
> 2 differences I noticed from the wireframes - not sure if it's intentional:
> 
> 1. The wireframe has the Premium Membership tab laid out more vertically:
> 
> Screen Shot 2022-05-05 at 3 08 43 pm
> 
> Compare to the current layout:
> 
> Screen Shot 2022-05-05 at 3 08 36 pm
> 
> 2. You have a divider line under your headings, whereas the Figma doesn't - also visible in the screenshots above. The reason I noticed is that it starts to carve up the page a little especially in Billing History:
> 
> Screen Shot 2022-05-05 at 3 12 32 pm
> 
> If design says it's fine then it's fine, just checking.

1. This is my fault overall - but since this wasn't mentioned at all in the task breakdown, I completely overlooked the new layout, but I'm also not sure its intentional as this is pretty far from what's in place today. I will check-in with design.

2. Fixed this - I made the changes while the builds were broken and wasn't able to test it until now. I've updated the related `sass` to remove dividers for tabbed pages and also adjusted the margins to closer resemble the wireframes.

File: src/app/settings/settings.component.ts
Comment (vincentsalucci) [reply src/app/settings/settings.component.ts:56]: Updated the screenshots to the latest changes for you @eliykat

**Comment by vincentsalucci** [comment]
@eliykat Talked with design - they've updated the wireframes to match what we currently have in place.

File: src/app/settings/user-billing-history.component.ts
Comment (Hinton) [inline src/app/settings/user-billing-history.component.ts:17]: We should really look into adopting https://angular.io/api/router/Resolve, these loading page behavior is bugging me.

File: src/app/settings/user-billing-history.component.ts
Comment (Hinton) [inline src/app/settings/user-billing-history.component.ts:32]: Feels like this belongs better in a guard class of some sort. Creating a tech debt item to figure out this kind of redirects.

File: src/app/organizations/settings/organization-billing.component.ts
Comment (Hinton) [inline src/app/organizations/settings/organization-billing.component.ts:56]: Can we enter this component without an orgnaizationId?

File: src/app/settings/payment-method.component.html
Comment (Hinton) [inline src/app/settings/payment-method.component.html:54]: We're duplicating this in two places, we should add a tech debt item to create a component for this.

File: src/app/settings/premium.component.html
Comment (Hinton) [inline src/app/settings/premium.component.html:4]: We are using a different pattern here than in `src/app/settings/user-subscription.component.html`. We should pick one.

**Review by Hinton — CHANGES_REQUESTED** [review]

**Review by Hinton — APPROVED** [review]
Approving this, assuming that we tackle the comments in a tech debt item afterwards.

File: src/app/settings/payment-method.component.html
Comment (vincentsalucci) [reply src/app/settings/payment-method.component.html:54]: This code duplication is only temporary and will be addressed with the Admin Vault Refresh.

File: src/app/settings/premium.component.html
Comment (vincentsalucci) [reply src/app/settings/premium.component.html:4]: That's because of the pre-existing loading pattern that's in place for the `user-subscription` component. I didn't feel that it was within scope to try and restructure that for the sole purpose of supplying a header text.

File: src/app/organizations/settings/organization-billing.component.ts
Comment (vincentsalucci) [reply src/app/organizations/settings/organization-billing.component.ts:56]: After reviewing the routing module/permissions for this component, it appears the answer to your question is no. This is just leftover for temporarily porting over the super class until the admin vault refresh.

Comment (Hinton) [reply src/app/organizations/settings/organization-billing.component.ts:56]: @vincentsalucci can you ensure we have a ticket for to ensure this gets completed in the admin refresh?

File: src/app/settings/premium.component.html
Comment (Hinton) [reply src/app/settings/premium.component.html:4]: @vincentsalucci can you ensure we have a tech debt ticket for this in the SG pod?

**Pull request merged**
